src/internal/utils/bucketHelper.ts

```diff
@@ -43,31 +42,21 @@ import {
 import { UICONFIG } from "../constants/uiConfig";
 import { getTranslation } from "./translations";
 
-export function removeUnusedFilters(filters: IFiltersBucketItem[], unusedBucketItems: IBucketItem[]) {
-    return filters.filter(filter => {
-        return (
-            filter.autoCreated === false ||
-            !unusedBucketItems.some(item => item.attribute === filter.attribute)
-        );
-    });
-}
-
-export function sanitizeUnusedFilters(
-    newReferencePoint: IExtendedReferencePoint,
-    referencePoint: IReferencePoint,
-): IExtendedReferencePoint {
-    const allAttributeBucketItems = getAllAttributeItems(referencePoint.buckets);
+export function sanitizeFilters(newReferencePoint: IExtendedReferencePoint): IExtendedReferencePoint {
     const usedAttributeBucketItems = getAllAttributeItems(newReferencePoint.buckets);
 
-    const unusedAttributeBucketItems = allAttributeBucketItems.filter(
-        (bucketItem: IBucketItem) =>
-            !usedAttributeBucketItems.some((used: IBucketItem) => used.attribute === bucketItem.attribute),
-    );
-
-    const filteredFilters = removeUnusedFilters(
-        get(newReferencePoint.filters, "items", []),
-        unusedAttributeBucketItems,
-    );
+    const filteredFilters = newReferencePoint.filters.items.filter((filterBucketItem: IFiltersBucketItem) => {
+        const hasCorrespondingBucketItem = usedAttributeBucketItems.some(
+            (attributeBucketItem: IBucketItem) => {
+                const filter = filterBucketItem.filters[0];
+                if (isAttributeFilter(filter) || isDateFilter(filter)) {
+                    return attributeBucketItem.attribute === filter.attribute;
+                }
+                return false;
+            },
+        );
+        return filterBucketItem.autoCreated === false || hasCorrespondingBucketItem;
```

Comment:
This way, it is way more readable than before. I would have one optimization suggestion, however. How about this way?

````
const filteredFilters = newReferencePoint.filters.items.filter((filterBucketItem: IFiltersBucketItem) => {
    if (filterBucketItem.autoCreated === false)  {
        return true;
    }
    return usedAttributeBucketItems.some(
        (attributeBucketItem: IBucketItem) => {
            const filter = filterBucketItem.filters[0];
            if (isAttributeFilter(filter) || isDateFilter(filter)) {
                return attributeBucketItem.attribute === filter.attribute;
            }
            return false;
        },
    );
});
````

Most of the filters are not auto-created and we could skip the computation of corresponding buckets in most of the cases.